RFC: Service Builder Improvements (#1620)
awslabs/smithy-rs: design/src/rfcs/rfc0020_service_builder.md

# RFC: Service Builder Improvements

> Status: RFC

> One might characterize `smithy-rs` as a tool for transforming a [Smithy service](https://awslabs.github.io/smithy/1.0/spec/core/model.html#service) into a [tower::Service](https://docs.rs/tower-service/latest/tower_service/trait.Service.html) builder. A Smithy model defines semantics and constrains the behavior of the generated service, but not completely - closures must be passed to the builder to before the `tower::Service` is fully specified. This builder structure is the primary API surface we provide to the customer, as a result, it is important that it meets their needs.

Author: It'd be nice to have a term which encompasses everything that implements `FnOnce`.

- **Smithy Service**: The entry point of an API that aggregates resources and operations together within a Smithy model. Described in detail [here](https://awslabs.github.io/smithy/1.0/spec/core/model.html#service).
- **Service**: The `tower::Service` trait is an interface for writing network applications in a modular and reusable way. `Service`s act on requests to produce responses.
- **Service Builder**: A `tower::Service` builder, generated from a Smithy service, by `smithy-rs`.
- **Middleware**: Broadly speaking, middleware modify requests and responses. Concretely, these are exist as implementations of [Layer](https://docs.rs/tower/latest/tower/layer/trait.Layer.html)/a `Service` wrapping an inner `Service`.

## Background

To provide context for the proposal we perform a survey of the current state of affairs.

The following is a reference model we will use throughout the RFC:

```smithy
operation Operation0 {
    input: Input0,
    output: Output0
}

operation Operation1 {
    input: Input1,
    output: Output1
}

service Service {
    operations: [
        Operation0,

>         Operation0,

Reviewer: `Operation1`

    ]
}
```

We have purposely omitted details from the model that are unimportant to describing the proposal. We also omit distracting details from the Rust snippets.

### Handlers

> An core concept in the current service builder is the `Handler` trait:

Reviewer 2: ```suggestion
A core concept in the current service builder is the `Handler` trait:
```

```rust
impl<Fun, Fut> Handler<(), Operation0Input> for Fun
where
    Fun: FnOnce(crate::input::Operation0Input) -> Fut,
    Fut: Future<Output = Operation0Output>,
{
    async fn call(self, request: http::Request) -> http::Response {
        let input = /* Create `Operation0Input` from `request: http::Request` */;

        // Use closure on the input
        let output = self(input).await;

        let response = /* Create `http::Response` from `output: Operation0Output` */
        response
    }
}

impl<Fun, Fut> Handler<Extension<S>, Operation0Input> for Fun
where
    Fun: FnOnce(Operation0Input, Extension<S>) -> Fut,
    Fut: Future<Output = Operation0Output>,
{
    async fn call(self, request: http::Request) -> http::Response {
        let input = /* Create `Operation0Input` from `request: http::Request` */;

        // Use closure on the input and fetched extension data
        let extension = Extension(request.extensions().get::<T>().clone());
        let output = self(input, extension).await;

        let response = /* Create `http::Response` from `output: Operation0Output` */
        response
    }
}
```

> Creating `Operation0Input` from a `http::Request` and `http::Response` from a `Operation0Output` involves the HTTP binding traits and protocol aware deserialization. The structure [RuntimeError](https://github.com/awslabs/smithy-rs/blob/458eeb63b95e6e1e26de0858457adbc0b39cbe4e/rust-runtime/aws-smithy-http-server/src/runtime_error.rs#L53-L5) enumerates internal error cases such as serialization/deserialization failures, `extensions().get::<T>()` failures, etc. We omit error handling in the snippets above, but, in full, they also involve protocol aware conversions from the `RuntimeError` and `http::Response`. The reader should make note of the influence of the model/protocol on the different sections of this procedure.

Reviewer 2: ```suggestion
Creating `Operation0Input` from a `http::Request` and `http::Response` from a `Operation0Output` involves the [HTTP binding traits](https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html) and protocol aware deserialization. The structure [RuntimeError](https://github.com/awslabs/smithy-rs/blob/458eeb63b95e6e1e26de0858457adbc0b39cbe4e/rust-runtime/aws-smithy-http-server/src/runtime_error.rs#L53-L5) enumerates error cases such as serialization/deserialization failures, `extensions().get::<T>()` failures, etc. We omit error handling in the snippets above, but, in full, they also involve protocol aware conversions from the `RuntimeError` and `http::Response`. The reader should make note of the influence of the model/protocol on the different sections of this procedure.
```

Despite not being `pub`, `RuntimeError` is not internal in that it gets exposed to the clients over the wire.

Author: I've worded this too roughly. What I'm trying to get at is that there are protocol specific failures which occur outside of the handler which can get captured by `RuntimeError`. What I'm going to argue for later is that, perhaps if we are going to have protocol specific routers, then perhaps it makes sense to have protocol specific errors (which all enjoy something akin to a `IntoResponse` trait).

> The `request.extensions().get::<T>()` present in the `Fun: FnOnce(Operation0Input, Extension<S>) -> Fut` implementation is the current approach to injecting state into handlers. The customer is required to apply a [AddExtensionLayer](https://docs.rs/tower-http/latest/tower_http/add_extension/struct.AddExtensionLayer.html) to the output of the service builder so that, when the request reaches the handler, the `extensions().get::<T>()` will succeed.

Reviewer 2: I'd call out in this paragraph that otherwise, the application panics, to better support your proposal.

Reviewer 2: My comment only applies to the second paragraph; I must have mistakenly selected both in the UI when writing it.

Reviewer 2: Actually, is this RFC going to address this extensions "type unsafety" issue? I thought you wanted to fix it but I don't see a relevant heading under the Proposal section.

Author: Will add a section about :+1:

To convert the closures described above into a `Service` `OperationHandler` is used:

```rust
pub struct OperationHandler<H, T, Input> {
    handler: H,
    _marker: PhantomData<(T, Input)>,
}

impl<H, T, Input> Service<Request<B>> for OperationHandler<H, T, Input>
where
    H: Handler<T, I>,
{
    type Response = http::Response;
    type Error = Infallible;

    #[inline]
    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    async fn call(&mut self, req: Request<B>) -> Result<Self::Response, Self::Error> {
        self.handler.call(req).await.map(Ok)
    }
}
```

> The service build uses both `Handler` and `OperationHandler`

Reviewer 2: ```suggestion
The service build uses both `Handler` and `OperationHandler`.
```

```rust
pub struct Route {
    service: BoxCloneService<http::Request, http::Response, Infallible>,
}

enum Routes {
    RestXml(Vec<(Route, RequestSpec)>),
    RestJson1(Vec<(Route, RequestSpec)>),
    AwsJson10(TinyMap<String, Route>),
    AwsJson11(TinyMap<String, Route>),
}

pub struct Router {
    routes: Routes,
}
```

and enjoys the following `Service<http::Request>` implementation:

```rust
impl Service<http::Request> for Router
{
    type Response = http::Response;
    type Error = Infallible;

    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    async fn call(&mut self, request: http::Request) -> Result<Self::Response, Self::Error> {
        match &self.routes {
            Routes::/* protocol */(routes) => {
                let route: Result<Route, _> = /* perform route matching logic */;
                match route {
                    Ok(ok) => ok.oneshot().await,
                    Err(err) => /* Convert routing error into http::Response */
                }
            }
        }
    }
}
```

> Along side the protocol specific constructors, `Router` includes a `layer` method. This provides a way for the customer to apply a `tower::Layer` to all routes. For every protocol, `Router::layer`, has the approximately the same behavior:

Reviewer 2: ```suggestion
Along side the protocol specific constructors, `Router` includes a `layer` method. This provides a way for the customer to apply a `tower::Layer` to all routes. For every protocol, `Router::layer` has the approximately the same behavior:
```

        Ok(OperationRegistry {
            operation0: self.operation0.ok_or(/* OperationRegistryBuilderError */)?,
            operation1: self.operation1.ok_or(/* OperationRegistryBuilderError */)?,
        })
    }
```

We can do away with fallibility if we allow for on `Op0`, `Op1`, to switch types during build. For example,

```rust
impl OperationRegistryBuilder<Op0, In0, Op1, In1> {
    pub fn operation0<NewOp0>(mut self, value: NewOp0) -> OperationRegistryBuilder<NewOp0, In0, Op1, In1> {
        OperationRegistryBuilder {
            operation0: value,
            operation1: self.operation1
        }
    }
    pub fn operation1<NewOp1>(mut self, value: NewOp1) -> OperationRegistryBuilder<Op0, In0, NewOp1, In1> {
        OperationRegistryBuilder {
            operation0: self.operation0,
            operation1: value
        }
    }
}

impl OperationRegistryBuilder<Op0, In0, Op1, In1>
where
    Op0: Handler<B, In0, Operation0Input>,
    Op1: Handler<B, In1, Operation1Input>,
{
    pub fn build(self) -> OperationRegistry<Op0, In0, Op1, In1> {
        OperationRegistry {
            operation0: self.operation0,

>             operation1: self.operation1,

Reviewer 3: I am a bit confused. Doesn't we get a type mismatch with `self.operation0` (an `Option`) and `OperationRegistry.operation0`?

Let me guess. Are you going to propose a solution in type-level state machine?

Author: Yep, that's exactly what the intention is here. I should mention that we get rid of `Option`s in addition to simply changing the signature of the setter methods.

Any solution which provides an `{Operation}` structure and wishes it to be accepted by multiple service builders must deal with this problem. We currently build one library per service and hence have duplicate structures when [service closures](https://awslabs.github.io/smithy/1.0/spec/core/model.html?highlight=closure#service-closure) overlap. This means we wouldn't run into this problem today, but it would be a future obstruction if we wanted to reduce the amount of generated code.

```rust
use tower::layer::util::{Stack, Identity};
use tower::util::{ServiceFn, service_fn};

// This takes the same form as `Operation0` defined in the previous attempt. The difference being that this is now
// private.
struct Service0Operation0<S> {
    inner: S
}

impl<S> Service<http::Request> for ServiceOperation0<S>
where
    S: Service<Operation0Input, Response = Operation0Output, Error = Infallible>
{
    /* Same as above */
}

pub struct Operation0<S, L> {
    inner: S,
    layer: L
}

impl<S> Operation0<S, Identity> {
    pub fn from_service(inner: S) -> Self {
        Self {
            inner,
            layer: Identity
        }
    }
}

impl<F> Operation0<ServiceFn<F>, Identity> {

>     pub fn from_handler(inner: F) -> Self {

Reviewer 2: Perhaps the omission is justified by the sentence above that said you're not going to deal with the whole handlers with and without state issue (or answered by the item in approach B "- `from_handler`, which takes an async `Operation0Input -> Operation0Output`."), but: in both this approach and approach B, what are the bounds on `F`? Is the idea to still require `F` implement `Handler`, which we will implement in codegen like we do now, for functions with and without state?

Author: Yep, we should continue with `Handler` pattern except the bounds change now. https://github.com/hlbarber/service-builder/pull/1/files#diff-67196527c762aa94fd3d640b4495ff46ab8e9a0760f426b9314b0e00b3f0846b

        Operation0::from_service(service_fn(inner))
    }
}

impl<S, L> Operation0<S, L> {
    pub fn layer<NewL>(self, layer: L) -> Operation0<S, Stack<L, NewL>> {
        Operation0 {
            inner: self.inner,
            layer: Stack::new(self.layer, layer)
        }
    }

    pub fn logging(self, /* args */) -> Operation0<S, Stack<L, LoggingLayer>> {
        Operation0 {
            inner: self.inner,
            layer: Stack::new(self.layer, LoggingLayer::new(/* args */))
        }
    }

    pub fn auth(self, /* args */) -> Operation0<S, Stack<L, AuthLayer>> {
        Operation0 {
            inner: self.inner,
            layer: Stack::new(self.layer, /* Construct auth middleware */)
        }

    }
}

impl<Op1, Op2> OperationRegistryBuilder<Op1, Op2> {

>     pub fn operation0<S, L>(self, operation: Operation0<S, L>) -> OperationRegistryBuilder<<L as Layer<Service0Operation0<S>>::Service, Op2>

Reviewer 2: I like approach C.

Perhaps I'm asking too much, but wouldn't it be nice if we could make `operation0` take in _both_ an `Operation0` struct, and a function/closure just as it does now? So `registry.operation0(fun)` is just syntax sugar for `registry.operation0(Operation0::from_handler(fun))`. Maybe by making `operation0` take a generic param implementing a trait that we auto-implement in codegen for both `Operation0` and the correct function types (with and without state)? Essentially an `UberHandler`.

Author: I think this is probably definitely doable and desirable. We could have `operation0` take a `Into<Operation<S, L>>` and implement `From<F>` where `F: Fn(/* with or without state */)` perhaps.

One thing I'm struggling with is that there's a lot of moving parts in this section and it's hard to maintain a narrative without side stepping. I'll make sure to get this in though.

Author: On second thought I don't think it's possible. It's possible to `impl Handler` for `F: Fn(X) -> ...` and `F: Fn(X, Y) -> ...` because it's sugar for `Fn<(X, )>` and `Fn<(X, Y)>` - no struct can be both so we can't run into conflicts. This isn't the case for `F: Fn` and `S: Service<Req>` - structs could presumably be both.

Author: On second thought I'm not sure this is possible because a `S: Service<Input, Response = Output>` could also be an async `Input -> Output`, in which case there would be conflicting conversion implementations.

        })
    }
```

We can do away with fallibility if we allow for on `Op0`, `Op1` to switch types during build and remove the `Option` from around the fields. The `OperationRegistryBuilder` then becomes

```rust
struct OperationRegistryBuilder<Op0, Op1> {
    operation_0: Op0,
    operation_1: Op1
}

impl OperationRegistryBuilder<Op0, In0, Op1, In1> {
    pub fn operation0<NewOp0>(mut self, value: NewOp0) -> OperationRegistryBuilder<NewOp0, In0, Op1, In1> {
        OperationRegistryBuilder {
            operation0: value,
            operation1: self.operation1
        }
    }
    pub fn operation1<NewOp1>(mut self, value: NewOp1) -> OperationRegistryBuilder<Op0, In0, NewOp1, In1> {
        OperationRegistryBuilder {
            operation0: self.operation0,
            operation1: value
        }
    }
}

impl OperationRegistryBuilder<Op0, In0, Op1, In1>
where
    Op0: Handler<B, In0, Operation0Input>,
    Op1: Handler<B, In1, Operation1Input>,
{
    pub fn build(self) -> OperationRegistry<Op0, In0, Op1, In1> {
        OperationRegistry {
            operation0: self.operation0,
            operation1: self.operation1,
        }
    }
}
```

> The customer will now get a compile time error rather than a runtime error when they fail to specify a handler.

Author: I think `required` would definitely be a good candidate for compile time checking.

The decision is not as as clear cut as the service builder case though:

- We already have generics on the service builder which allows us to type check.
- "missing handler" is the entire failure criteria so we eliminate fallibility completely.

> The customer will now get a compile time error rather than a runtime error when they fail to specify a handler.

Reviewer 2: Thinking more about it, I think it's not worth introducing API asymmetry for `required` with respect to the other constraint traits. Maybe it'd be weird to have one constraint trait checked at compile time and the others at runtime. And modifying the deserializers to make them aware of when they've set a value would be a hassle.

```rust
let user_routes = Router::new().route("/:id", /* service */);

let team_routes = Router::new().route("/", /* service */);

let api_routes = Router::new()
    .nest("/users", user_routes)
    .nest("/teams", team_routes);

let app = Router::new().nest("/api", api_routes);
```

Note that, in `axum` handlers are quickly converted to `tower::Service` (via `IntoService`) before they are passed into the `Router`. In contrast, in `smithy-rs` handlers are passed into a builder and then the conversion to `tower::Service` is performed (via `OperationHandler`).

Introducing state to handlers in `axum` is done in the same way as `smithy-rs`, described briefly in [Handlers](#handlers) - a layer is used to insert state into incoming `http::Request`s and the `Handler` implementation pops it out of the type map layer. In `axum`, if a customer wanted to scope state to all routes within `/users/` they are able to do the following:

```rust
async fn handler(Extension(state): Extension</* State */>) -> /* Return Type */ {}

let api_routes = Router::new()
    .nest("/users", user_routes.layer(Extension(/* state */)))
    .nest("/teams", team_routes);
```

> In `smithy-rs` a customer is only able to apply a layer to either the `aws_smithy_http::routing::Router` or every route via the [layer method](#router) described above.

Author: The `Router` itself is a `Service` so the customer can wrap it in middleware.

> In `smithy-rs` a customer is only able to apply a layer to either the `aws_smithy_http::routing::Router` or every route via the [layer method](#router) described above.

Reviewer 2: But currently that's the only place one can apply a layer in smithy-rs?

# Proposal

The proposal is presented as a series of compatible transforms to the existing service builder, each paired with a motivation. Most of these can be independently implemented, but in the case where there exists an interdependency it is stated.

> Although presented as a mutation to the existing service builder, the actual implementation should exist as a entirely separate builder - reusing code generation from the old builder but exposes a new Rust API. Preserving the old API surface will prevent breakage and make it easier to perform comparative benchmarks and testing.

Reviewer 2: Is this goal achievable for _all_ proposed transformations? E.g. How do we expose an entirely separate `OperationRegistryBuilder` without changing the type's name(space)?

> ## Remove two-step build procedure

Reviewer 2: I assume this transformation will be superseded by the yet-to-be-drafted "Type erasure with the name of the Smithy service" (?) They are both independent, but since they're also both breaking changes, I'd just release one of them. If this is the case I'd note it down somewhere in the RFC.

        response
    }
}

fn make_operation_0<S>(inner: S) -> Operation0<S> {
    Operation0 {
        inner
    }
}
```

which is concerned with all the same model aware serialization/deserialization we noted in [Handler](#handler). The API usage then becomes

```rust
async fn handler0(input: Operation0Input) -> Operation0Output {
    todo!()
}

async fn handler1(input: Operation1Input, state: State) -> Operation1Output {
    todo!()
}

// Model agnostic `into_service` method which transforms a handler `{Operation}Input -> {Operation}Output` to
// `tower::Service<{Operation}Input, Response = {Operation}Output, Error = Infallible>`
let svc = OperationHandler::new(handler0);

// Middleware can be applied at this point
let svc = model_layer.layer(svc);

// Model aware conversion between `tower::Service<{Operation}Input, Response = {Operation}Output, Error = Infallible>`
// and `tower::Service<http::Request, Response = http::Response, Error = Infallible>`
let operation0 = make_operation0(svc);

// Middleware can be applied at this point
let operation0 = http_layer.layer(op1_svc);

OperationRegistryBuilder::default()

>     .operation0(operation0)

Reviewer 2: What is preventing the user from passing in an arbitrary `tower::Service` for `operation0` here? The current operation registry ensures that the provided handler takes in and returns the appropriate input and output types.

Author: Nothing, it's possible that the customer can pass a `tower::Service<http::Request, Response = http::Response>` here which completely ignores the model. I'm unsure if this is any worse than allowing the customer to apply arbitrary `Layer`s.

Approach C has been added which is more attractive from this POV.
